payjoin/src/psbt/mod.rs

```diff
@@ -182,41 +184,56 @@ impl InternalInputPair<'_> {
-                    Some(script) if script.is_witness_program() && script.is_p2wpkh() =>
-                        Ok(NESTED_P2WPKH_MAX),
-                    // Other script or witness program.
-                    Some(_) => Err(InputWeightError::NotSupported),
-                    // No redeem script provided. Cannot determine the script type.
-                    None => Err(InputWeightError::NoRedeemScript),
-                }
-            }
-            P2wpkh => Ok(InputWeightPrediction::P2WPKH_MAX),
-            P2wsh => Err(InputWeightError::NotSupported),
-            P2tr => Ok(InputWeightPrediction::P2TR_KEY_DEFAULT_SIGHASH),
-            _ => Err(AddressTypeError::UnknownAddressType.into()),
-        }?;
-
-        // Lengths of txid, index and sequence: (32, 4, 4).
-        let input_weight = iwp.weight() + Weight::from_non_witness_data_size(32 + 4 + 4);
-        Ok(input_weight)
+        expected_input_weight(self.address_type()?, self)
     }
 }
 
+pub fn expected_input_weight(
+    address_type: AddressType,
+    internal_input_pair: &InternalInputPair,
+) -> Result<Weight, InputWeightError> {
+    use bitcoin::AddressType::*;
+
+    let (psbtin, txin) = (internal_input_pair.psbtin, internal_input_pair.txin);
+
+    // Get the input weight prediction corresponding to spending an output of this address type
+    let iwp = match address_type {
+        P2pkh => Ok(InputWeightPrediction::P2PKH_COMPRESSED_MAX),
+        P2sh => {
+            // redeemScript can be extracted from scriptSig for signed P2SH inputs
+            let redeem_script = if let Some(ref script_sig) = psbtin.final_script_sig {
+                script_sig.redeem_script()
+                // try the PSBT redeem_script field for unsigned inputs.
+            } else {
+                psbtin.redeem_script.as_ref().map(|script| script.as_ref())
+            };
+            match redeem_script {
+                // Nested segwit p2wpkh.
+                Some(script) if script.is_witness_program() && script.is_p2wpkh() =>
+                    Ok(NESTED_P2WPKH_MAX),
+                // Other script or witness program.
+                Some(_) => Err(InputWeightError::NotSupported),
+                // No redeem script provided. Cannot determine the script type.
+                None => Err(InputWeightError::NoRedeemScript),
+            }
+        }
+        P2wpkh => Ok(InputWeightPrediction::P2WPKH_MAX),
+        P2wsh =>
+            if psbtin.final_script_witness.is_some() {
+                let witness_size = txin.segwit_weight().to_wu() as usize;
+                Ok(InputWeightPrediction::new(0, [witness_size]))
+            } else {
+                Err(InputWeightError::NotSupported)
+            },
+        P2tr => Ok(InputWeightPrediction::P2TR_KEY_DEFAULT_SIGHASH),
+        _ => Err(AddressTypeError::UnknownAddressType.into()),
+    }?;
+
+    // Add the non_witness weight.
+    let input_weight = iwp.weight() + NON_WITNESS_INPUT_WEIGHT;
```

Comment:
I agree the const is better but I would leave this unchanged. If we are going to change it lets do it in a seperate commit
`let input_weight = iwp.weight() + Weight::from_non_witness_data_size(32 + 4 + 4);`